mycroft/util/process_utils.py

```diff
@@ -121,3 +122,131 @@ def echo(message):
+        on_started (callable): callback to execute when process is started.
+        on_alive (callable): callback to execute when minimum operation is
+                             reached.
+        on_ready (callable): callback to execute when loading is 100 %
+                             complete.
+        on_stopping (callable): callback to execute when shutdown of process
+                                starts.
+        on_error (callable): callback to execute when an unrecoverable error
+                             renders the process unusable.
+    """
+    def __init__(self, name, bus,
+                 on_started=None, on_alive=None, on_ready=None,
+                 on_stopping=None, on_error=None):
+
+        # Messagebus connection
+        self.bus = bus
+        self.name = name
+
+        # Callback functions
+        self.on_started = on_started
+        self.on_alive = on_alive
+        self.on_ready = on_ready
+        self.on_error = on_error
+        self.on_stopping = on_stopping
+
+        self.state = ProcessState.NOT_STARTED
+        self._register_handlers()
+
+    def _register_handlers(self):
+        """Register messagebus handlers for status queries."""
+        self.bus.on('mycroft.{}.is_alive'.format(self.name), self.check_alive)
+        self.bus.on('mycroft.{}.is_ready'.format(self.name),
+                    self.check_ready)
+        self.bus.on('mycroft.{}.all_loaded'.format(self.name),
+                    self.check_ready)
+
+    def check_alive(self, message=None):
+        """Respond to is_alive status request.
+
+        Arguments:
+            message: Optional message to respond to, if omitted no message
+                     is sent.
+
+        Returns:
+            bool, True if process is alive.
+        """
+        is_alive = self.state >= ProcessState.ALIVE
+
+        if message:
+            status = {'status': is_alive}
+            self.bus.emit(message.response(data=status))
+
+        return is_alive
+
+    def check_ready(self, message=None):
+        """Respond to all_loaded status request.
+
+        Arguments:
+            message: Optional message to respond to, if omitted no message
+                     is sent.
+
+        Returns:
+            bool, True if process is ready.
+        """
+        is_ready = self.state >= ProcessState.READY
+        if message:
+            status = {'status': is_ready}
+            self.bus.emit(message.response(data=status))
+
+        return is_ready
+
+    def set_started(self):
+        """Process is started."""
+        self.state = ProcessState.STARTED
+        if self.on_started:
+            self.on_started()
+
+    def set_alive(self):
+        """Basic loading is done."""
+        self.state = ProcessState.ALIVE
+        if self.on_alive:
+            self.on_alive()
+
+    def set_ready(self):
+        """All loading is done."""
+        self.state = ProcessState.READY
+        if self.on_ready:
+            self.on_ready()
```

Comment:
This will differ wildly between the different processes. In skills "basic loading" is defined as priority skills are loaded and the process has started it's main sequence but all skills aren't yet loaded.

In the audio service basic loading is when TTS has been loaded and mycroft can start talking, (servicing the priority skills for example), all loading is when the audioservices are loaded.